### PR TITLE
Pipeline Topology Implementation

### PR DESCRIPTION
**Pull request by andrewballantyne**
**Fixes**: 
https://issues.redhat.com/browse/ODC-2605

**Analysis / Root cause**: 
Implements the foundation of the Pipeline Builder and replaces the existing rendering logic to be more inline with what the Pipeline spec can present as a node network.

**Solution Description**: 
Pulled the `@console/topology` code into Pipelines. This gives us the power of a layout engine and a free and open canvas to draw in.

**Screen shots / Gifs for design review**: 
@openshift/team-devconsole-ux 
Screen Shot 2020-01-18 at 7 37 34 PM

Note: the top visualization is the same rendering issue from https://issues.redhat.com/browse/ODC-2464 fixed.

There are some unexpected outcomes out of the layout as of right now:

Screen Shot 2020-01-18 at 7 37 45 PM

I have no current intent on addressing the above issues until after the Pipeline Builder feature is completed. We'll likely need to work on a constraint system with Dagre which may use "ghost" groups or something (unsure of the solution at this time).

Pipeline Runs:

PipelineTopology-ComplexPipelineRun

PipelineTopology-P2P

**Unit test coverage report**: 
- [ ] TODO: Add tests

**Test setup:**
 - Install the OpenShift Pipeline Operator
 - Add pipelines (could install the ones from `frontend/packages/dev-console/yamls/pipelines/install_pipeline_mocks.sh` or create / upload your own)
 - View the new visualization in Pipeline and Pipeline Runs (and in the Add flow)

**Browser conformance**: 
- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

/kind feature
cc @jeff-phillips-18 @christianvogt

### CODE REVIEW

File: frontend/packages/dev-console/src/components/pipelines/detail-page-tabs/pipeline-details/PipelineVisualization.tsx [hunk 1]
```diff
@@ -1,15 +1,33 @@
...
+const PipelineVisualization: React.FC<PipelineTopologyVisualizationProps> = ({
+  pipeline,
+  pipelineRun,
+}) => {
+  const { nodes, edges } = getTopologyNodesEdges(pipeline, pipelineRun);
+
+  if (nodes.length === 0 && edges.length === 0) {
+    // Nothing to render
+    // TODO: Confirm wording with UX; ODC-1860
+    return <Alert variant="info" isInline title="This Pipeline has no tasks to visualize." />;
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/components/pipelines/detail-page-tabs/pipeline-details/PipelineVisualization.tsx]: @serenamarie125 I took a causal stab at solving https://issues.redhat.com/browse/ODC-1860 when I stumbled upon the issue in my work.

File: frontend/packages/dev-console/src/components/pipelines/detail-page-tabs/pipeline-details/PipelineVisualizationTask.scss [hunk 2]
```diff
@@ -2,7 +2,6 @@ $border-color: var(--pf-global--BorderColor--light-100);
 .odc-pipeline-vis-task {
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/components/pipelines/detail-page-tabs/pipeline-details/PipelineVisualizationTask.scss]: I likely can gut a lot of this too... I just haven't looked too much into it.

File: frontend/packages/dev-console/src/components/pipelines/detail-page-tabs/pipeline-details/pipeline-topology/factories.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        default:
+          return undefined;
+      }
+    default:
+      return undefined;
+  }
+};
+
+// TODO: Fix this hack as it's not the best way to get the layout update
+type CallbackLayout = (onLayout: LayoutCallback) => LayoutFactory;
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/components/pipelines/detail-page-tabs/pipeline-details/pipeline-topology/factories.ts]: @jeff-phillips-18 This is the hack we discussed - we need to find a better way for me to get a notification when the layout finishes.

File: frontend/packages/topology/src/layouts/DagreLayout.ts [hunk 4]
```diff
@@ -69,13 +76,23 @@ class DagreEdge {
   }
 }
 
+export type LayoutCallback = (nodes: Node[], edges: Edge[]) => void;
+type AdditionalDagreLayoutOptions = { onLayout: LayoutCallback };
 export default class DagreLayout implements Layout {
```
Comment (andrewballantyne) [inline frontend/packages/topology/src/layouts/DagreLayout.ts]: All changes for `DagreLayout` were rushed and not really thought through. @jeff-phillips-18 @christianvogt Feel free to comment on direction / changes. I'll likely conflict with your work Jeff when it comes to your topology work.

**Comment by serenamarie125** [comment]
@andrewballantyne this is looking really good!  will this PR also cover the tooltips associated with the PLR visualization?  I'm assuming they shouldn't be affected, but you never know :) 

Is it also worth including screenshots of what the PL vs PLR visualizations look like?

**Comment by andrewballantyne** [comment]
> @andrewballantyne this is looking really good! will this PR also cover the tooltips associated with the PLR visualization? I'm assuming they shouldn't be affected, but you never know :)
> 
> Is it also worth including screenshots of what the PL vs PLR visualizations look like?

Doh, of course, I'll update with it. I tested it under the hood but didn't post the screenshots 😦 Not sure what I was thinking haha.

**Comment by andrewballantyne** [comment]
@serenamarie125 Apologizes for that; Pipeline Runs have been added to the description (from the Pipeline Run details page).

**Comment by spadgett** [comment]
/hold
this didn't actually pass e2e tests

**Comment by spadgett** [comment]
/hold cancel
rerunning tests

File: frontend/packages/dev-console/src/components/pipelines/pipeline-topology/TaskEdge.tsx [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  const targetNode = element.getTarget();
+
+  return (
+    <path
+      d={integralShapePath(
+        startPoint.clone().translate(sourceNode.getBounds().width / 2 - 1, 0),
+        endPoint.clone().translate(-targetNode.getBounds().width / 2, 0),
+      )}
+      stroke="var(--pf-global--BorderColor--light-100)"
+      fill="none"
```
Comment (jeff-phillips-18) [inline frontend/packages/dev-console/src/components/pipelines/pipeline-topology/TaskEdge.tsx]: Rather these styles be in a `.scss` file but ok.

File: frontend/packages/topology/src/layouts/DagreLayout.ts (hunk 4 above)
Comment (jeff-phillips-18) [reply frontend/packages/topology/src/layouts/DagreLayout.ts]: We can reconcile once the changes are in, you can use the `GRAPH_LAYOUT_END_EVENT` event once its there.

**Comment by andrewballantyne** [comment]
Folded into #4055 for CI sanity.

**Pull request closed without merging**